core/services/wifi/WifiManager.py

```diff
@@ -113,78 +84,125 @@ def __dict_from_list(data: bytes) -> Dict[str, Any]:
 
         return output
 
-    def get_wifi_available(self) -> List[ScannedWifiNetwork]:
+    async def get_wifi_available(self) -> List[ScannedWifiNetwork]:
         """Get a dict from the wifi signals available"""
-        try:
-            scan_data = self.wpa.send_command_scan()
-            if b"BUSY" in scan_data:
-                raise BusyError
-            data = self.wpa.send_command_scan_results()
-            networks_list = WifiManager.__dict_from_table(data)
-            return [ScannedWifiNetwork(**network) for network in networks_list]
-        except BusyError as error:
-            raise BusyError("WPA service is busy. Try to space network scans by at least 4 seconds.") from error
-        except Exception as error:
-            raise FetchError("Failed to fetch wifi list.") from error
 
-    def get_saved_wifi_network(self) -> List[SavedWifiNetwork]:
+        async def perform_new_scan() -> None:
+            try:
+                self._scan_task = asyncio.create_task(self.wpa.send_command_scan(timeout=15))
+                await self._scan_task
+                data = await self.wpa.send_command_scan_results()
+                networks_list = WifiManager.__dict_from_table(data)
+                self._updated_scan_results = [ScannedWifiNetwork(**network) for network in networks_list]
+            except Exception as error:
+                if self._scan_task is not None:
+                    self._scan_task.cancel()
+                self._updated_scan_results = None
+                raise FetchError("Failed to fetch wifi list.") from error
+
+        # Performs a new scan only if it's the first one or the last one is already done
+        # In case there's one running already, wait for it to finish and use its result
+        if self._scan_task is None:
+            await perform_new_scan()
+        elif self._scan_task.done():
+            await perform_new_scan()
```

Comment:
This should work:
```suggestion
        if self._scan_task is None or self._scan_task.done():
            await perform_new_scan()
```
The second clause should only run if the first one is false:
```
>>> batata = None
>>> if batata is None or batata.biscoito():
...     print("wat")
... 
wat

```

Reply:
This one passed by.
Changed.